src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts

```diff
@@ -139,19 +140,36 @@ export class TerminalInstance implements ITerminalInstance {
 		this._initDimensions();
 		this._createProcess(this._shellLaunchConfig);
 		this._createXterm();
+		if (this._shellLaunchConfig) {
+			this.setTitle(path.basename(this._shellLaunchConfig.executable), true);
+		}
 
 		if (platform.isWindows) {
 			this._processReady.then(() => {
 				this._windowsShellHelper = new WindowsShellHelper(this._processId, this._shellLaunchConfig.executable);
+				this._onCheckWindowsShell = new Emitter<TPromise<string>>();
+				// The debounce is necessary to prevent multiple processes from spawning when
+				// the enter key or output is spammed
+				debounceEvent(this._onCheckWindowsShell.event, (l, e) => e, 200, true)(() => {
+					this.checkWindowShell();
+				});
+				this._xterm.on('lineFeed', () => this._onCheckWindowsShell.fire());
+				this._xterm.on('keypress', () => this._onCheckWindowsShell.fire());
```

Comment:
If we're doing this here we still need the enter check and the `this._messageTitleListener` check.